scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryService.scala

```diff
@@ -1,9 +1,798 @@
+                // Try to bond back, if this is a new node.
+                bond(caller)
+              )
+              .startAndForget
+            // Return the latet local ENR sequence.
+            enrSeq <- localEnrSeq
+          } yield Some(Some(enrSeq))
+
+    /** Handle incoming FindNode request. */
+    override def findNode: Call[Peer[A], Proc.FindNode] =
+      caller =>
+        target =>
+          respondIfBonded(caller, "FindNode") {
+            for {
+              state <- stateRef.get
+              targetId = Node.kademliaId(target)
+              closestNodeIds = state.kBuckets.closestNodes(targetId, config.kademliaBucketSize)
+              closestNodes = closestNodeIds
+                .map(state.kademliaIdToNodeId)
+                .map(state.nodeMap)
+            } yield closestNodes
+          }
+
+    /** Handle incoming ENRRequest. */
+    override def enrRequest: Call[Peer[A], Proc.ENRRequest] =
+      caller => _ => respondIfBonded(caller, "ENRRequest")(stateRef.get.map(_.enr))
+
+    // The methods below are `protected[v4]` so that they can be called from tests individually.
+    // Initially they were in the companion object as pure functions but there are just too many
+    // parameters to pass around.
+
+    protected[v4] def toPeer(node: Node): Peer[A] =
+      Peer(node.id, toAddress(node.address))
+
+    protected[v4] def currentTimeMillis: Task[Long] =
+      clock.realTime(MILLISECONDS)
+
+    protected[v4] def localEnrSeq: Task[ENRSeq] =
+      stateRef.get.map(_.enr.content.seq)
+
+    /** Check if the given peer has a valid bond at the moment. */
+    protected[v4] def isBonded(
+        peer: Peer[A]
+    ): Task[Boolean] = {
+      currentTimeMillis.flatMap { now =>
+        stateRef.get.map { state =>
+          if (state.isSelf(peer))
+            true
+          else
+            state.lastPongTimestampMap.get(peer) match {
+              case None =>
+                false
+              case Some(timestamp) =>
+                timestamp > now - config.bondExpiration.toMillis
+            }
+        }
+      }
+    }
+
+    /** Return Some response if the peer is bonded or log some hint about what was requested and return None. */
+    protected[v4] def respondIfBonded[T](caller: Peer[A], request: String)(response: Task[T]): Task[Option[T]] =
+      isBonded(caller).flatMap {
+        case true => response.map(Some(_))
+        case false => Task(logger.debug(s"Ignoring $request request from unbonded $caller")).as(None)
+      }
+
+    /** Runs the bonding process with the peer, unless already bonded.
+      *
+      * If the process is already running it waits for the result of that,
+      * it doesn't send another ping.
+      *
+      * If the peer responds it waits for a potential ping to arrive from them,
+      * so we can have some reassurance that the peer is also bonded with us
+      * and will not ignore our messages.
+      */
+    protected[v4] def bond(
```

Comment:
from what i understand bonding is `ping - pong - fetch_enr` sequence ?

Reply:
Strictly speaking it just tries to achieve a mutual ping, so want to have reasonable confidence that this happened, if it has to:
```
Ping -> 
        <- Pong
        <- Ping
Pong ->
```
It's possible that the other side didn't need to Ping us, because we have responded earlier to a Pong, in the last 12 hours.